kratos/python_scripts/analysis_stage.py

```diff
@@ -1,305 +1,384 @@
+
+        for process in self._GetListOfProcesses():
+            process.ExecuteFinalizeSolutionStep()
+
+    def OutputSolutionStep(self):
+        """This function printed / writes output files after the solution of a step
+        """
+        # first we check if one of the output processes will print output in this step
+        # this is done to save computation in case none of them will print
+        is_output_step = False
+        for output_process in self._GetListOfOutputProcesses():
+            if output_process.IsOutputStep():
+                is_output_step = True
+                break
+
+        if is_output_step: # at least one of the output processes will print output
+            for process in self._GetListOfProcesses():
+                process.ExecuteBeforeOutputStep()
+
+            for output_process in self._GetListOfOutputProcesses():
+                if output_process.IsOutputStep():
+                    output_process.PrintOutput()
+
+            for process in self._GetListOfProcesses():
+                process.ExecuteAfterOutputStep()
+
+    def Check(self):
+        """This function checks the AnalysisStage
+        """
+        self._GetSolver().Check()
+        for process in self._GetListOfProcesses():
+            process.Check()
+
+    def Clear(self):
+        """This function clears the AnalysisStage
+        """
+        self._GetSolver().Clear()
+
+    def ModifyInitialProperties(self):
+        """this is the place to eventually modify material properties in the stage """
+        pass
+
+    def ModifyInitialGeometry(self):
+        """this is the place to eventually modify geometry (for example moving nodes) in the stage """
+        pass
+
+    def ModifyAfterSolverInitialize(self):
+        """this is the place to eventually do any modification that requires the solver to be initialized """
+        pass
+
+    def ApplyBoundaryConditions(self):
+        """here the boundary conditions is applied, by calling the InitializeSolutionStep function of the processes"""
+
+        for process in self._GetListOfProcesses():
+            process.ExecuteInitializeSolutionStep()
+
+        #other operations as needed
+
+    def ChangeMaterialProperties(self):
+        """this function is where the user could change material parameters as a part of the solution step """
+        pass
+
+    def _GetSolver(self):
+        if not hasattr(self, '_solver'):
+            self._solver = self._CreateSolver()
+        return self._solver
+
+    def _CreateSolver(self):
+        """Create the solver
+        """
+        raise Exception("Creation of the solver must be implemented in the derived class.")
+
+    ### Modelers
+    def _SetupGeometryModelAndModelParts(self):
+        ## Initialize Modelers
+        self._list_of_modelers = self._CreateModelers(self._GetOrderOfModelers())
+
+        # Import or generate geometry models from external input.
+        for modeler in self._GetListOfModelers():

```

Comment:
I would have put each of this as a function.

aside the debugging echo is not included